compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -2537,4 +2564,160 @@ public int hashCode() {
+     * one element
+     *
+     * @param type BFiniteType union or finite
+     * @return boolean whether type has a valid filler value or not
+     */
+    private boolean checkFillerValue(BFiniteType type) {
+        if (type.isNullable()) {
+            return true;
+        }
+
+        // For singleton types, that value is the implicit initial value
+        if (type.getValueSpace().size() == 1) {
+            return true;
+        }
+
+        boolean defaultFillValuePresent = false;
+
+        Iterator iterator = type.getValueSpace().iterator();
+        BLangExpression firstElement = (BLangExpression) iterator.next();
+        BType firstElementType = firstElement.type;
+        String defaultFillValue = getDefaultFillValue(firstElement);
+        if (firstElement.toString().equals(defaultFillValue)) {
+            defaultFillValuePresent = true;
+        }
+
+        while (iterator.hasNext()) {
+            Object value =  iterator.next();
+            BType valueType = ((BLangExpression) value).type;
+            if (!isSameType(valueType, firstElementType)) {
+                return false;
+            }
+            if (!defaultFillValuePresent && value.toString().equals(defaultFillValue)) {
+                defaultFillValuePresent = true;
+            }
+        }
+        return defaultFillValuePresent;
+    }
+
+    private boolean checkFillerValue(BUnionType type) {
+        if (type.isNullable()) {
+            return true;
+        }
+        Iterator<BType> iterator = type.getMemberTypes().iterator();
+        BType firstMember = iterator.next();
+        boolean defaultFillValuePresent = false;
+
+        // is first value is a valid fill value
+        String defaultFillValue = getDefaultFillValue(firstMember);
+        if (firstMember.toString().equals(defaultFillValue)) {
+            defaultFillValuePresent = true;
+        }
+
+        while (iterator.hasNext()) {
+            Object value = iterator.next();
+
+            if (!isSameType(firstMember, iterator.next())) {
+                return false;
+            }
+
+            if (!defaultFillValuePresent && value.toString().equals(defaultFillValue)) {
+                defaultFillValuePresent = true;
+            }
+        }
+        return defaultFillValuePresent;
+    }
+
+    private boolean checkFillerValue(BRecordType type) {
+        for (BField field : type.fields) {
+            if (Symbols.isFlagOn(field.symbol.flags, Flags.OPTIONAL)) {
+                continue;
+            }
+            if (Symbols.isFlagOn(field.symbol.flags, Flags.REQUIRED)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean checkFillerValue(BArrayType type) {
+        return hasFillerValue(type.eType);
+    }
+
+    private String getDefaultFillValue(BLangExpression expr) {
+        return getDefaultFillValue(expr.type);
+    }
+
+    private String getDefaultFillValue(BType type) {
```

Comment:
Wondering if keeping this in the String format is the best approach? Can't we use the value itself and compare, based on the assumption that they will only be used for comparison with either literals or constant refs?

Reply:
If the integer is '5' the valid fill value should be '0' for an int. Need to get the default value or implicit default value for int type which is zero. 
I have added a negative test case for this.
Test Case:  https://github.com/ballerina-platform/ballerina-lang/pull/20749/files#diff-65f99a2264d6173185afec0d71e151c6R63
Error Message: https://github.com/ballerina-platform/ballerina-lang/pull/20749/files#diff-7319ccb85e364b3233fb96f11c424f29R212

Reply:
Note: The code is written similar to run-time check. https://github.com/ballerina-platform/ballerina-lang/blob/7e0d4677ed81725d69b1a01c52df2a47333589dd/bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java#L1787

The plan is to replace string enum with objects and also in accordance with spec. This may require calling initialiser methods to create zero value objects for every type.